### DISCUSSION TITLE
ATtiny1624: 140uA in SLEEP_MODE_PWR_DOWN

### DISCUSSION THREAD

**Discussion by MagSem** [SpenceKonde/megaTinyCore#502] [Library discussion]
Hi.
I have read all issues, both open and closed.
I took a new chip ATtiny1624 (only a capacitor for power supply) and flashed a simple program.

#include <avr/sleep.h>
#include <avr/power.h>

void PORT_Init(void)
{
  pinMode(PIN_PA0, INPUT_PULLUP);
  pinMode(PIN_PA1, INPUT_PULLUP);
  pinMode(PIN_PA2, INPUT_PULLUP);
  pinMode(PIN_PA3, INPUT_PULLUP);
  pinMode(PIN_PA4, INPUT_PULLUP);
  pinMode(PIN_PA5, INPUT_PULLUP);
  pinMode(PIN_PA6, INPUT_PULLUP);
  pinMode(PIN_PA7, INPUT_PULLUP);
  pinMode(PIN_PB0, INPUT_PULLUP);
  pinMode(PIN_PB1, INPUT_PULLUP);
  pinMode(PIN_PB2, INPUT_PULLUP);
  pinMode(PIN_PB3, INPUT_PULLUP);
}

void setup()
{
  PORT_Init();
  power_all_disable();
  set_sleep_mode(SLEEP_MODE_PWR_DOWN);
  sleep_enable();
}
 
void loop()
{
  sleep_cpu();
}

The current consumption (1 MHz, 3 V, all off) was 140 μA.

I read that on chips of the first series, some people managed to achieve 100 nA, but in ATtiny1624 I could not get such a value.
Are there people who have reached 100 nA on chips of the second series?
Maybe the new ADC won't turn off?

Regards,
MagSem

**Comment by SpenceKonde** [comment]
#include <avr/power.h>

and all macros in it which you call all have no effect on any modern AVR. If you work your way through power.h, you'll see that everything in the body of hte macros is all #ifdef''ed out. 

I haven't explored sleep on these parts yet, too many higher priorities but take a look through the spec'ed expectatioins for current consumption. specifically the peripherals. ADC sucks 260uA at 1MHz, and megaTinyCore clocks it at around 2.5x that, so it's not the ADC, it'd be over half a mA if it was that!

The oscillator is right about in the right ballpark, except it definitely sholdn't be on in power down sleep mode.... OH!!

Did you program it over UPDI? Try giving it a full cold powercycle (like, disconnect suppply, and then short vcc and ground together. then reconnect power (after removing the short obviously - that step is because sometimes capacitors retain just enough charge to prevent a clean restart on what seems like you'd be powercycling something.  If UPDI doesn't properly get disabled after programming it won't let the oscillator shut off. I didn't look at that here because I assumed that was handled correctly by Microchips PyMCUProg code that we started from.

**Reply by MagSem** [reply]
Yes, over UPDI. Without a capacitor - the same thing.
When replacing sleep_cpu()  with while(1), the current consumption is 618 μA, as in the specification.
Something does not want to fall asleep.
I also use an STM8L050J3 chip, so it is possible to reach 400 nA there, carefully disconnecting all the peripherals in the IAR.
In short, without switching to the ultra-low current consumption mode, it is impossible to fully use this chip.
In general, I would like to thank you for the work done.

**Reply by mechatroniks-git** [reply]
> If UPDI doesn't properly get disabled after programming it won't let the oscillator shut off. I didn't look at that here because I assumed that was handled correctly by Microchips PyMCUProg code that we started from.

Is there a UPDI command we can send to disable, or do I write a 1 to the UPDIDIS bit on UPDI.CTRLB on bootup?

```31.3.2.2.2 UPDI Regular Disable
Any programming or debugging session that does not require any specific operation from the UPDI after
disconnecting the programmer has to be terminated by writing the UPDI Disable (UPDIDIS) bit in the Control B
(UPDI.CTRLB) register, upon which the UPDI will issue a System Reset and disable itself. The Reset will restore the
CPU to the Run state, independent of the previous state. It will also lower the UPDI clock request to the system and
reset any UPDI KEYs and settings.
If the disable operation is not performed, the UPDI and the oscillator’s request will remain enabled. This causes
increased power consumption for the application.
```

```
Bit 2 – UPDIDIS UPDI Disable
Writing a ‘1’ to this bit disables the UPDI PHY interface. The clock request from the UPDI is lowered, and the UPDI is
reset. All the UPDI PHY configurations and keys will be reset when the UPDI is disabled.
```

**Comment by MagSem** [comment]
I found the problem.
If you use only `main `(as C style) instead of `setup `and `loop `(as Arduino style), then the current consumption actually drops to 0.1 - 0.2 μA.
It is worth returning the `setup `and `loop` - again 140 μA.
Regards,
MagSem

**Comment by mechatroniks-git** [comment]
I tried this, it appears to run as a simple blink, no idea how to measure draw in the microamps range. Do I need one of [these](https://www.nordicsemi.com/Products/Development-hardware/Power-Profiler-Kit-2), or is there a better option?

```
#include <avr/sleep.h>

void RTC_init(void)
{
  /* Initialize RTC: */
  while (RTC.STATUS > 0)
  {
    ;                                   /* Wait for all register to be synchronized */
  }
  RTC.CLKSEL = RTC_CLKSEL_INT32K_gc;    /* 32.768kHz Internal Ultra-Low-Power Oscillator (OSCULP32K) */

  RTC.PITINTCTRL = RTC_PI_bm;           /* PIT Interrupt: enabled */

  RTC.PITCTRLA = RTC_PERIOD_CYC16384_gc /* RTC Clock Cycles 16384, resulting in 32.768kHz/16384 = 2Hz */
  | RTC_PITEN_bm;                       /* Enable PIT counter: enabled */
}

ISR(RTC_PIT_vect)
{
  RTC.PITINTFLAGS = RTC_PI_bm;          /* Clear interrupt flag by writing '1' (required) */
}

void initialize() {
  init();
  RTC_init();                           /* Initialize the RTC timer */
  pinMode(PIN_PA0, INPUT_PULLUP);
  pinMode(PIN_PA1, INPUT_PULLUP);
  pinMode(PIN_PA2, INPUT_PULLUP);
  pinMode(PIN_PA3, INPUT_PULLUP);
  pinMode(PIN_PA4, INPUT_PULLUP);
  pinMode(PIN_PA5, INPUT_PULLUP);
  pinMode(PIN_PA6, OUTPUT);             /* Configure pin PIN_PA6 as an output */
  pinMode(PIN_PA7, INPUT_PULLUP);
  pinMode(PIN_PB0, INPUT_PULLUP);
  pinMode(PIN_PB1, INPUT_PULLUP);
  pinMode(PIN_PB2, INPUT_PULLUP);
  pinMode(PIN_PB3, INPUT_PULLUP);
  set_sleep_mode(SLEEP_MODE_PWR_DOWN);  /* Set sleep mode to POWER DOWN mode */
  sleep_enable();                       /* Enable sleep mode, but not going to sleep yet */
}

int main() {
  initialize();
  while(1){
    sleep_cpu();                          /* Sleep the device and wait for an interrupt to continue */
    digitalWrite(PIN_PA6, CHANGE);        /* Device woke up and toggle LED on pin#7 */
  }
  return(0);
}
```

**Reply by MagSem** [reply]
Here are the measurement results of your example with ADS1262 (32-bit ADC with calibr resistor) and LabVIEW, **140 uA** in lowpower.

![Sleep_1](https://user-images.githubusercontent.com/46042293/130191799-5c856f3b-6152-4f22-aa37-c6f76c254a3a.png)

But I'm talking a little about something else. Let's look at an example with a button interrupt, not RTC.

```
#include <avr/io.h> 
#include <avr/interrupt.h>
#include <avr/sleep.h>
#include <util/delay.h>

#define PB3_INTERRUPT             PORTB.INTFLAGS & PIN3_bm
#define PB3_CLEAR_INTERRUPT_FLAG  PORTB.INTFLAGS &= PIN3_bm

//------------------------------------------------------------------------>
// PB3_INTERRUPT ISR
//------------------------------------------------------------------------>
ISR(PORTB_PORT_vect)
{
  if(PB3_INTERRUPT)                 // PB3 changed state
  {
    digitalWrite(PIN_PA3, CHANGE);  // toggle the LED
    PB3_CLEAR_INTERRUPT_FLAG;
  }
}
//------------------------------------------------------------------------<

//------------------------------------------------------------------------>
// Set pins to Low Power Mode
//------------------------------------------------------------------------>
void PORT_LOWPOWER_Init(void)
{
  for (uint8_t i = 0; i < 8; i++)
  {
    *((uint8_t *)&PORTA + 0x10 + i) |= 1 << PORT_PULLUPEN_bp;
  }
  for (uint8_t i = 0; i < 8; i++)
  {
    *((uint8_t *)&PORTB + 0x10 + i) |= 1 << PORT_PULLUPEN_bp;
  }
}
//------------------------------------------------------------------------<

//------------------------------------------------------------------------>
// Set pins to Working Mode
//------------------------------------------------------------------------>
void PORT_WORKING__Init(void)
{
  // Enable PA3 for output
  PORTA.DIR = PIN3_bm;

  // Enable PB3 for input
  PORTB.DIRCLR = PIN3_bm;
  PORTB.PIN3CTRL = PORT_PULLUPEN_bm;
  PORTB.PIN3CTRL |= PORT_ISC_BOTHEDGES_gc;
}
//------------------------------------------------------------------------<

/*
void setup()
{
  PORT_LOWPOWER_Init();
  PORT_WORKING__Init();

  // Pause for debugging/measuring
  PORTA.OUTTGL = PIN3_bm;
  _delay_ms(1000);
  PORTA.OUTTGL = PIN3_bm;
  
  // Enable interrupts
  sei();

  set_sleep_mode(SLEEP_MODE_PWR_DOWN);
  sleep_enable();
}
 
void loop()
{
  while (1)
  {
    sleep_cpu();
  }
}
*/

///*
int main(void)
{
  PORT_LOWPOWER_Init();
  PORT_WORKING__Init();

  // Pause for debugging/measuring
  PORTA.OUTTGL = PIN3_bm;
  _delay_ms(1000);
  PORTA.OUTTGL = PIN3_bm;
  
  // Enable interrupts
  sei();

  set_sleep_mode(SLEEP_MODE_PWR_DOWN);
  sleep_enable();

  while (1)
  {
    sleep_cpu();
  }
}
//*/

```
If we use `setup()-loop()`, the result is as follows (key press), **140 uA** in lowpower:

![Sleep_2](https://user-images.githubusercontent.com/46042293/130193562-eca72597-9799-4605-abb7-887875878c64.png)

If we use `main()-while(1)`, the result is as follows (key press), **0,17 uA** in lowpower:

![Sleep_3](https://user-images.githubusercontent.com/46042293/130193975-bf87d816-aeef-428e-bcf5-f3e41faec1f8.png)

As for the measuring equipment, there are a huge number of options, a multimeter with a μA range is enough.

Regards,
MagSem

**Comment by mechatroniks-git** [comment]
If you compile both versions and compare the .lst files (mine is at C:\Users\w10\AppData\Local\Temp\arduino_build_393584/test.ino.lst)

The setup/loop .lst file has a bunch of extra code (from wiring.c?), where it looks like ADC, interrupts and timers are being configured? This code is missing from the main() .lst file. 


```
main():
C:\Users\win10\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.3.2\cores\megatinycore/wiring.c:679
    #elif (F_CPU == 16000000)
      /* No division on clock */
      _PROTECTED_WRITE(CLKCTRL_MCLKCTRLB, 0x00);
    #elif (F_CPU == 10000000) //20MHz prescaled by 2
      /* Clock DIV2 */
      _PROTECTED_WRITE(CLKCTRL_MCLKCTRLB, (CLKCTRL_PEN_bm | CLKCTRL_PDIV_2X_gc));
 15a:	81 e0       	ldi	r24, 0x01	; 1
 15c:	98 ed       	ldi	r25, 0xD8	; 216
 15e:	94 bf       	out	0x34, r25	; 52
 160:	80 93 61 00 	sts	0x0061, r24	; 0x800061 <__RODATA_PM_OFFSET__+0x7f8061>
init_ADC0():
C:\Users\win10\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.3.2\cores\megatinycore/wiring.c:729
    #if F_CPU     > 24000000    // 24 MHz / 16 = 1.5 MHz,  25 MHz / 32 =  780 kHz
      ADC0.CTRLC  = ADC_PRESC_DIV16_gc | ADC_REFSEL_VDDREF_gc | ADC_SAMPCAP_bm;
    #elif F_CPU  >= 12000000    // 16 MHz / 16 = 1.0 MHz,  20 MHz / 16 = 1.25 MHz
      ADC0.CTRLC  = ADC_PRESC_DIV16_gc | ADC_REFSEL_VDDREF_gc | ADC_SAMPCAP_bm;
    #elif F_CPU  >=  6000000    //  8 MHz /  8 = 1.0 MHz,  10 MHz /  8 = 1.25 MHz
      ADC0.CTRLC  = ADC_PRESC_DIV8_gc  | ADC_REFSEL_VDDREF_gc | ADC_SAMPCAP_bm;
 164:	92 e5       	ldi	r25, 0x52	; 82
 166:	90 93 02 06 	sts	0x0602, r25	; 0x800602 <__RODATA_PM_OFFSET__+0x7f8602>
C:\Users\win10\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.3.2\cores\megatinycore/wiring.c:735
    #elif F_CPU  >=  3000000    //  4 MHz /  4 = 1.0 MHz,   5 MHz /  4 = 1.25 MHz
      ADC0.CTRLC  = ADC_PRESC_DIV4_gc  | ADC_REFSEL_VDDREF_gc | ADC_SAMPCAP_bm;
    #else                       //  1 MHz /  2 = 500 kHz - the lowest setting
      ADC0.CTRLC  = ADC_PRESC_DIV2_gc  | ADC_REFSEL_VDDREF_gc | ADC_SAMPCAP_bm;
    #endif
    ADC0.SAMPCTRL = (16); // 16 ADC clocks, 16 us
 16a:	90 e1       	ldi	r25, 0x10	; 16
 16c:	90 93 05 06 	sts	0x0605, r25	; 0x800605 <__RODATA_PM_OFFSET__+0x7f8605>
C:\Users\win10\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.3.2\cores\megatinycore/wiring.c:736
    ADC0.CTRLD    = ADC_INITDLY_DLY16_gc;
 170:	90 e2       	ldi	r25, 0x20	; 32
 172:	90 93 03 06 	sts	0x0603, r25	; 0x800603 <__RODATA_PM_OFFSET__+0x7f8603>
C:\Users\win10\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.3.2\cores\megatinycore/wiring.c:737
    ADC0.CTRLA   |= ADC_ENABLE_bm;
 176:	90 91 00 06 	lds	r25, 0x0600	; 0x800600 <__RODATA_PM_OFFSET__+0x7f8600>
 17a:	91 60       	ori	r25, 0x01	; 1
 17c:	90 93 00 06 	sts	0x0600, r25	; 0x800600 <__RODATA_PM_OFFSET__+0x7f8600>
init_TCA0():
C:\Users\win10\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.3.2\cores\megatinycore/wiring.c:816

  /* PORTMUX setting for TCA - don't need to set because using default */
  //PORTMUX.CTRLA = PORTMUX_TCA00_DEFAULT_gc;

  /* Enable Split Mode */
  TCA0.SPLIT.CTRLD = TCA_SPLIT_SPLITM_bm;
 180:	80 93 03 0a 	sts	0x0A03, r24	; 0x800a03 <__RODATA_PM_OFFSET__+0x7f8a03>
C:\Users\win10\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.3.2\cores\megatinycore/wiring.c:821

  //Only 1 WGM so no need to specifically set up.

  /* Period setting, 8-bit register in SPLIT mode */
  TCA0.SPLIT.LPER    = PWM_TIMER_PERIOD;
 184:	9e ef       	ldi	r25, 0xFE	; 254
 186:	90 93 26 0a 	sts	0x0A26, r25	; 0x800a26 <__RODATA_PM_OFFSET__+0x7f8a26>
C:\Users\win10\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.3.2\cores\megatinycore/wiring.c:822
  TCA0.SPLIT.HPER    = PWM_TIMER_PERIOD;
 18a:	90 93 27 0a 	sts	0x0A27, r25	; 0x800a27 <__RODATA_PM_OFFSET__+0x7f8a27>
C:\Users\win10\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.3.2\cores\megatinycore/wiring.c:839

  /* Use prescale appropriate for system clock speed */
  #if (F_CPU > 25000000) //   use 256 divider when clocked over 25 MHz - probably not terribly relevant - though they might be viable at 30 or 32, and are viable at 24/25 MHz.
    TCA0.SPLIT.CTRLA   = (TCA_SPLIT_CLKSEL_DIV256_gc) | (TCA_SPLIT_ENABLE_bm);
  #elif (F_CPU > 5000000) //  use 64 divider
    TCA0.SPLIT.CTRLA   =  (TCA_SPLIT_CLKSEL_DIV64_gc) | (TCA_SPLIT_ENABLE_bm);
 18e:	9b e0       	ldi	r25, 0x0B	; 11
 190:	90 93 00 0a 	sts	0x0A00, r25	; 0x800a00 <__RODATA_PM_OFFSET__+0x7f8a00>
init_millis():
C:\Users\win10\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.3.2\cores\megatinycore/wiring.c:617
        //RTC.CLKSEL=0; this is the power on value
      #endif
      RTC.INTCTRL         = 0x01; //enable overflow interrupt
      RTC.CTRLA           = (RTC_RUNSTDBY_bm|RTC_RTCEN_bm|RTC_PRESCALER_DIV32_gc);//fire it up, prescale by 32.
    #else //It's a type b timer - we have already errored out if that wasn't defined
      _timer->CCMP = TIME_TRACKING_TIMER_PERIOD;
 194:	27 e8       	ldi	r18, 0x87	; 135
 196:	33 e1       	ldi	r19, 0x13	; 19
 198:	20 93 5c 0a 	sts	0x0A5C, r18	; 0x800a5c <__RODATA_PM_OFFSET__+0x7f8a5c>
 19c:	30 93 5d 0a 	sts	0x0A5D, r19	; 0x800a5d <__RODATA_PM_OFFSET__+0x7f8a5d>
C:\Users\win10\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.3.2\cores\megatinycore/wiring.c:619
      // Enable timer interrupt, but clear the rest of register
      _timer->INTCTRL = TCB_CAPT_bm;
 1a0:	80 93 55 0a 	sts	0x0A55, r24	; 0x800a55 <__RODATA_PM_OFFSET__+0x7f8a55>
C:\Users\win10\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.3.2\cores\megatinycore/wiring.c:621
      // Clear timer mode (since it will have been set as PWM by init())
      _timer->CTRLB=0;
 1a4:	10 92 51 0a 	sts	0x0A51, r1	; 0x800a51 <__RODATA_PM_OFFSET__+0x7f8a51>
C:\Users\win10\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.3.2\cores\megatinycore/wiring.c:623
      // CLK_PER/1 is 0b00,. CLK_PER/2 is 0b01, so bitwise OR of valid divider with enable works
      _timer->CTRLA = TIME_TRACKING_TIMER_DIVIDER|TCB_ENABLE_bm;  // Keep this last before enabling interrupts to ensure tracking as accurate as possible
 1a8:	83 e0       	ldi	r24, 0x03	; 3
 1aa:	80 93 50 0a 	sts	0x0A50, r24	; 0x800a50 <__RODATA_PM_OFFSET__+0x7f8a50>
main():
C:\Users\win10\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.3.2\cores\megatinycore/wiring.c:662
  #ifndef MILLIS_USE_TIMERNONE
    init_millis();
  #endif
  /*************************** ENABLE GLOBAL INTERRUPTS *************************/
  // Finally, after everything is initialized, we go ahead and enable interrupts.
  sei();
 1ae:	78 94       	sei
 1b0:	e0 e1       	ldi	r30, 0x10	; 16
 1b2:	f4 e0       	ldi	r31, 0x04	; 4

```

**Comment by mechatroniks-git** [comment]
Look at this: https://github.com/SpenceKonde/megaTinyCore/issues/143#issuecomment-770442520

try adding
```
  power_all_disable();

```

**Reply by MagSem** [reply]
As SpenceKonde wrote above:

"#include <avr/power.h>

and all macros in it which you call all have no effect on any modern AVR. If you work your way through power.h, you'll see that everything in the body of hte macros is all #ifdef''ed out."

power_all_disable() in <avr/power.h> and I already tried using it.

At the moment we can try to disable all modules manually.

**Comment by mechatroniks-git** [comment]
@MagSem Can you try adding this to your setup/loop version to disable all the things being enabled, and measure the current?

``` 
  ADC0.CTRLA &= ~ADC_ENABLE_bm;
  TCD0.CTRLA = 0;
  TCA0.SPLIT.CTRLA = 0;
  RTC.CTRLA = 0;
```

**Reply by SpenceKonde** [reply]
I knew it would end up being something like this, there's just something that gets turned on that wew don't know was still running in sleep. I couldn't 

You can also overridee init functions with empty ones one at a time until you find the offender.... 

  init_clock(); 
  init_ADC0();
  init_timers();
  init_millis();
  
Though I stikll don't know how any of those should be m,aking it through to POWER DOWN sleep mode. Standby maybe.....

Do them in reverse order that they're on this list, as they have dependancies on the ones before them. 

Ooooooooh

Did you read the documentation in the readme about ADC power options, specificically

##### void ADCPowerOptions(options) *2-series only*
The PGA is enabled by any call to `analogReadEnh()` or `analogReadDiff()` that specifies valid gain > 0, by default we turn it off afterwards. There is also a "low latency" mode that, when enabled, keeps the ADC reference and related hardware running to prevent the delay (on order of tens of microseconds) before the next analog reading is taken). We use that by default, but it can be turned off with this function. The argument is a number between 0x00 and 0x0F. The structure is 0b 0000 PPLL - P bits control the PGA power state, L bits the LOWLAT setting.
* `LL = 00 or 01` Take no action on low latency mode.
* `LL = 10` Turn off low latency mode.
* `LL = 11` Turn on low latency mode.
* `PP = 00` Don't change anything relating to the PGA.
* `PP = 01` Turn off the PGA now. Don't change settings; if not set to turn off automatically, that doesn't change.
* `PP = 10` Enable PGA. Disable the automatic shutoff of the PGA.
* `PP = 11` Disable PGA now, and in future turn if off after use

**Reply by SpenceKonde** [reply]
Generally speaking, the core defaults to being optimized for performance, not power consumption by the initial configuration.  Moew power managmement stuff will be handlked as part of #158

**Comment by MagSem** [comment]
As expected, this is a new ADC.
With  `ADC0.CTRLA &= ~ADC_ENABLE_bm;` (or comment `init()` in main.cpp), with 3V from battery, with 1 MHz:

![Sleep_4](https://user-images.githubusercontent.com/46042293/130401031-ba37071d-5099-48d8-a69d-58a7276a1e7a.png)

Thanks for not making me look for this in the manual, @Mechatroniks )

@SpenceKonde, оf course, you, as the author, are free to make the kernel as you see fit (performance optimization, clarity and convenience for Arduino users, etc.), I just wanted to figure out if I could use this chip for my projects.
Thanks again for your work.

Regards,
MagSem

**Reply by mechatroniks-git** [reply]
Thanks for looking into this @MagSem, good to know the ADC is the culprit and a quick fix.

**Comment by SpenceKonde** [comment]
Bey specifically it's the low lat mode

____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore: Arduino support for almost every ATTiny microcontroller
Contact: ***@***.***

**Comment by dattasaurabh82** [comment]
Some study like [this](http://www.gammon.com.au/power) will be highly appreciated on this [page](https://github.com/SpenceKonde/megaTinyCore/blob/master/megaavr/extras/PowerSave.md) using information on methods from new ATTINYs to shed some more light.

**Comment by SpenceKonde** [comment]
Any study of sleep modes and power consumption goes under the purview of
#158, which I don't see happening for a few months; I need to get a patch
for some super critical issues out for DxCore so that I can get to work
ASAP getting ATTinyCore 2.0.0 out because a paying client needs that one
for during the upcoming school year. (words cannot describe how badly I
want to be able to declare ATTinyCore "maintenance only" so I can focus on
the new stuff. But like 90% of the work over there is done. Only once
that's done can I puit serious time into understanding the sleep and power
consumption behavior of tje modern AVR parts.(I recognize that
there's massive demand for it, but there's a critical bug on DxCore and a
paying client who needs the tiny26 support on ATTinyCore which is mostly
done but in the infrastructure iof 2.0.0 which currently doesn't have a a
boards.txt file (I think the only way to make that core viable (and really
all these cores, moving forward) is to stop with the hand edited boards.txt
- that is a huge loser and has caused more bugs than any other single area
of the dcore.  and so is completely unusable and untestable. I mean hell,
One of the things I need to fix in this week's panic 2.4.1 release
isthat while playing with something else, I realized I'd botched a clock
speed menu entry. Have you ever looked at boards.txt? It is a nightmare
format. I don't think I could design a format more difficult to hand edit
without introducing bugs than the boards.txt format is if I tried to do so.

On Tue, Aug 24, 2021 at 4:12 AM Saurabh Datta ***@***.***>
wrote:

> Some study like this <http://www.gammon.com.au/power> will be highly
> appreciated on this page
> <https://github.com/SpenceKonde/megaTinyCore/blob/master/megaavr/extras/PowerSave.md>
> using information on methods from new ATTINYs to shed some more light.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/SpenceKonde/megaTinyCore/discussions/502#discussioncomment-1226191>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABTXEW2GVCTSV35Z7EUHL4DT6NH7LANCNFSM5CJEM4FQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>


-- 
____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore <https://github.com/SpenceKonde/ATTinyCore>: Arduino support for
all pre-2016 tinyAVR with >2k flash!
megaTinyCore <https://github.com/SpenceKonde/megaTinyCore>: Arduino support
for all post-2016 tinyAVR parts!
DxCore <https://github.com/SpenceKonde/DxCore>: Arduino support for the AVR
Dx-series parts, the latest and greatest from Microchip!
Contact: ***@***.***

**Reply by dattasaurabh82** [reply]
I know someone from the Arduino team who shares your emotions and has been asking them internally to change few things ... But delayed as well. 😅

**Comment by SpenceKonde** [comment]
Oh good lord I don;'t want the Arduino team to do anything about this! Please! Those guys have a reverse midas touch!

**Comment by SpenceKonde** [comment]
I have more information than my last reply,. On the 2-series parts, the ADC
should be turned off before entering powerdown sleep (because we have to
keep lowlat mode on to work around serious errata) this costs just about
130 uA (similar to how classic AVR ADC burned 200uA at all times including
sleep if not disabled.

Disable the ADC. In 1.5.12, you can do that with analogPowerOptions(), but
that's not implemented in 1.5.11, the current released version, 1.5.12 is
almost released though

I think SerialUPDI automatically disables UPDI when it closes the
connection.

On Tue, Jul 26, 2022 at 5:01 PM Mechatroniks ***@***.***>
wrote:

> Did you program it over UPDI? Try giving it a full cold powercycle (like,
> disconnect suppply, and then short vcc and ground together. then reconnect
> power (after removing the short obviously - that step is because sometimes
> capacitors retain just enough charge to prevent a clean restart on what
> seems like you'd be powercycling something. If UPDI doesn't properly get
> disabled after programming it won't let the oscillator shut off. I didn't
> look at that here because I assumed that was handled correctly by
> Microchips PyMCUProg code that we started from.
>
> Is there a UPDI command we can send to disable, or do I write a 1 to the
> UPDIDIS bit on UPDI.CTRLB on bootup?
>
> Any programming or debugging session that does not require any specific operation from the UPDI after
>
> disconnecting the programmer has to be terminated by writing the UPDI Disable (UPDIDIS) bit in the Control B
>
> (UPDI.CTRLB) register, upon which the UPDI will issue a System Reset and disable itself. The Reset will restore the
>
> CPU to the Run state, independent of the previous state. It will also lower the UPDI clock request to the system and
>
> reset any UPDI KEYs and settings.
>
> If the disable operation is not performed, the UPDI and the oscillator’s request will remain enabled. This causes
>
> increased power consumption for the application.
>
> Bit 2 – UPDIDIS UPDI Disable
>
> Writing a ‘1’ to this bit disables the UPDI PHY interface. The clock request from the UPDI is lowered, and the UPDI is
>
> reset. All the UPDI PHY configurations and keys will be reset when the UPDI is disabled.
>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/SpenceKonde/megaTinyCore/discussions/502#discussioncomment-3255411>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABTXEWZILUTIWTXA6IXAOODVWBG2DANCNFSM5CJEM4FQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>


-- 
____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore <https://github.com/SpenceKonde/ATTinyCore>: Arduino support for
all pre-2016 tinyAVR with >2k flash!
megaTinyCore <https://github.com/SpenceKonde/megaTinyCore>: Arduino support
for all post-2016 tinyAVR parts!
DxCore <https://github.com/SpenceKonde/DxCore>: Arduino support for the AVR
Dx-series parts, the latest and greatest from Microchip!
Contact: ***@***.***

**Comment by InstantMuffin** [comment]
Not sure if this is a bug in the comparator library or if this should be noted in the docs or if I just missed anything.
But when using the comparator library and the comparator is active, you have to call Comparator.stop() before disabling the ADC like shown in the powerdown example for the 2-series, otherwise the MCU will lock itself up after waking up. I had a serial output directly after the wakeup (because I was playing around with powerdown and the AC and I thought disabling the ADC would somehow cause the AC to be disabled too in some way since I wasn't getting any outputs) and it just spewed out garbage before hanging up completely, so it seems the MCU does seem to continue, then do something unexpected, and then lock itself up.

**Reply by SpenceKonde** [reply]
That's for the heads up, Which parts does this effect? 2-series only ,or 0/1 roo

So comparator tooo,,,

**Reply by InstantMuffin** [reply]
I only have 1627s and 3227s and so far only tested and confirmed this to happen with the 1627.
I would guess that naturally the AC would need or share circuitry with the ADC, however I wouldn't expect this to behave that way and furthermore I could not find anything about this in the datasheet or the errata sheet.

If someone else can confirm this maybe we should leave a note in the powersave and comparator manuals (or errata), just so people know about this even though the powersave doc needs to be reworked anyway.

**Comment by mechatroniks-git** [comment]
Yeah, I'm calling `ADC0.CTRLA &= ~ADC_ENABLE_bm;  // disable adc` and can get well under my target of less than 30uA, but only if I pull power and short the cap after programming. 

I did a late-night firmware update without pulling the battery & shorting caps with a few boards, each board's battery died in about 10 days on all of them. I recharged, they have been holding at 100% battery for past 5 days. So I'm pretty sure it's the UPDI not disabling the oscillator. 

I'll try disabling with the register and report back.

**Reply by InstantMuffin** [reply]
Oh, FYI, I don't know about your code and configuration (including other components you maybe have connected), but I found that there is some current draw during powerdown if a serial port is active. At 3.3V I measured around 26µA. Just mentioning this since you say you are measuring around the same ballpark number.

**Reply by mechatroniks-git** [reply]
Thanks, yea, my max sleep power draw is related to other circuit devices on my pcb, the ATtiny will draw ~1uA when I have that isolated. 

Not using uart, but thanks for the tip on the AC needing to be stopped, I wasn't aware of that.

**Comment by SpenceKonde** [comment]
The /avr/power.h library should not be used on modern AVRs it has not been updated. Alll of the macros amount to dd nothings, because none of the registers they test for exist on moden AVR. Don't use that library. Axtually I think iI can make it warn if power.h ia included...

This adds some required features to he sleep and power fo sho (autooff+restore?)
But in the short term I will refer to tables 33-7 and 33-5. Between those two it is black and white clear as crystal that peripherals must be turnd off fefore entering sleep.

**Comment by mechatroniks-git** [comment]
Back to UPDIDIS - I was looking to test power draw, but not sure how to write this bit from my code. They are only readable from the UPDI programming interface. No mention on how to set it or if it's possible. Any tips on how to set this bit from my code?

31.5 Register Description
These registers are readable only through the UPDI with special instructions and are not readable through the CPU.
@SpenceKonde mentioned that the UPDI programmer should be setting this bit. I see this python code in MegaTinyCore, but after downloading my code, I also use pymcuprog to write some calibration stuff in the USERROW. Pymcuprog has the same block of code. No idea if either one is actually calling this code or if this code actually does set the UPDIDIS bit. Python is not my thing.

    def leave_progmode(self):
        """
        Disables UPDI which releases any keys enabled
        """
        self.logger.info("Leaving NVM programming mode")
        self.reset(apply_reset=True)
        self.reset(apply_reset=False)
        self.readwrite.write_cs(constants.UPDI_CS_CTRLB,
                                (1 << constants.UPDI_CTRLB_UPDIDIS_BIT) | (1 << constant



> Python is not my thing.


_It ain't my thing  nor the authors' of pymcuprog either. 

I'd test if that's the issue by power off, short the caps, and power back up without UPDIing and see fi there's a difference_

**Comment by mechatroniks-git** [comment]
Just ran a test, my board draws 5uA in sleep after UPDI. Same if I power down and short the caps.  The 5uA is leakage from the other circuits, which for my use is not an issue, the battery will self-discharge faster than 5uA drawdown will.
